Error after close DB client

https://github.com/couchbase/couchbase-lite-java/issues/52

```java
private void shutdownDatabase() {
        push.goOffline();
        pull.goOffline();
        pull.stop();
        push.stop();
        database.close();
        manager.close();
    }
````
in log:
```
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@32f4b68 rejected from java.util.concurrent.ScheduledThreadPoolExecutor@5263919f[Shutting down, pool size = 1, active threads = 1, queued tasks = 6, completed tasks = 15]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
	at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
	at java.util.concurrent.ScheduledThreadPoolExecutor.submit(ScheduledThreadPoolExecutor.java:632)
	at java.util.concurrent.Executors$DelegatedExecutorService.submit(Executors.java:678)
	at com.couchbase.lite.replicator.PullerInternal.changeTrackerStopped(PullerInternal.java:766)
	at com.couchbase.lite.replicator.ChangeTracker.stopped(ChangeTracker.java:512)
	at com.couchbase.lite.replicator.ChangeTracker.stop(ChangeTracker.java:498)
	at com.couchbase.lite.replicator.PullerInternal.goOffline(PullerInternal.java:994)
	at com.couchbase.lite.replicator.ReplicationInternal$15.doIt(ReplicationInternal.java:1101)
	at com.couchbase.lite.replicator.ReplicationInternal$15.doIt(ReplicationInternal.java:1098)
	at com.github.oxo42.stateless4j.StateConfiguration$3.doIt(StateConfiguration.java:130)
	at com.github.oxo42.stateless4j.StateConfiguration$3.doIt(StateConfiguration.java:127)
	at com.github.oxo42.stateless4j.StateRepresentation.executeEntryActions(StateRepresentation.java:123)
	at com.github.oxo42.stateless4j.StateRepresentation.enter(StateRepresentation.java:102)
	at com.github.oxo42.stateless4j.StateMachine.publicFire(StateMachine.java:199)
	at com.github.oxo42.stateless4j.StateMachine.fire(StateMachine.java:125)
	at com.couchbase.lite.replicator.ReplicationInternal$1.run(ReplicationInternal.java:191)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@4579bbfb rejected from java.util.concurrent.ScheduledThreadPoolExecutor@5263919f[Shutting down, pool size = 1, active threads = 1, queued tasks = 3, completed tasks = 18]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
	at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
	at java.util.concurrent.ScheduledThreadPoolExecutor.submit(ScheduledThreadPoolExecutor.java:632)
	at java.util.concurrent.Executors$DelegatedExecutorService.submit(Executors.java:678)
	at com.couchbase.lite.replicator.ReplicationInternal.notifyChangeListeners(ReplicationInternal.java:983)
	at com.couchbase.lite.replicator.ReplicationInternal.notifyChangeListenersStateTransition(ReplicationInternal.java:1138)
	at com.couchbase.lite.replicator.ReplicationInternal.access$400(ReplicationInternal.java:61)
	at com.couchbase.lite.replicator.ReplicationInternal$16.doIt(ReplicationInternal.java:1109)
	at com.couchbase.lite.replicator.ReplicationInternal$16.doIt(ReplicationInternal.java:1105)
	at com.github.oxo42.stateless4j.StateRepresentation.executeExitActions(StateRepresentation.java:130)
	at com.github.oxo42.stateless4j.StateRepresentation.exit(StateRepresentation.java:112)
	at com.github.oxo42.stateless4j.StateMachine.publicFire(StateMachine.java:197)
	at com.github.oxo42.stateless4j.StateMachine.fire(StateMachine.java:125)
	at com.couchbase.lite.replicator.ReplicationInternal$1.run(ReplicationInternal.java:191)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I'm sorry(((( if your class Test run with:
```Thread.sleep(1000 * 30);```
then app don't close, with log:
```
SEVERE: Test: PULL replication event. Source: com.couchbase.lite.replicator.Replication@7388716e Total changes: 15374 Completed changes: 14447
мар 12, 2015 12:31:59 PM com.couchbase.lite.util.SystemLogger e
SEVERE: Test: PULL replication event. Source: com.couchbase.lite.replicator.Replication@7388716e Total changes: 15375 Completed changes: 14447
мар 12, 2015 12:31:59 PM com.couchbase.lite.util.SystemLogger e
SEVERE: Test: PULL replication event. Source: com.couchbase.lite.replicator.Replication@7388716e Total changes: 15376 Completed changes: 14447
мар 12, 2015 12:31:59 PM com.couchbase.lite.util.SystemLogger e
SEVERE: Test: PULL replication event. Source: com.couchbase.lite.replicator.Replication@7388716e Total changes: 15377 Completed changes: 14447
мар 12, 2015 12:31:59 PM com.couchbase.lite.util.SystemLogger e
SEVERE: Sync: ERROR: processInbox failed: 
java.lang.NullPointerException
	at com.couchbase.lite.replicator.PullerInternal.processInbox(PullerInternal.java:158)
	at com.couchbase.lite.replicator.ReplicationInternal$2.process(ReplicationInternal.java:298)
	at com.couchbase.lite.support.Batcher.processNow(Batcher.java:207)
	at com.couchbase.lite.support.Batcher.access$100(Batcher.java:23)
	at com.couchbase.lite.support.Batcher$1.run(Batcher.java:40)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

мар 12, 2015 12:31:59 PM com.couchbase.lite.util.SystemLogger e
SEVERE: Batcher: com.couchbase.lite.support.Batcher$1@7595b28: BatchProcessor throw exception
java.lang.RuntimeException: java.lang.NullPointerException
	at com.couchbase.lite.replicator.ReplicationInternal$2.process(ReplicationInternal.java:302)
	at com.couchbase.lite.support.Batcher.processNow(Batcher.java:207)
	at com.couchbase.lite.support.Batcher.access$100(Batcher.java:23)
	at com.couchbase.lite.support.Batcher$1.run(Batcher.java:40)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at com.couchbase.lite.replicator.PullerInternal.processInbox(PullerInternal.java:158)
	at com.couchbase.lite.replicator.ReplicationInternal$2.process(ReplicationInternal.java:298)
	... 10 more

мар 12, 2015 12:31:59 PM com.couchbase.lite.util.SystemLogger e
SEVERE: Test: PULL replication event. Source: com.couchbase.lite.replicator.Replication@7388716e Transition: STOPPING -> STOPPED Total changes: 15377 Completed changes: 14447
THREAD 0: Thread[Signal Dispatcher,9,system]
THREAD 1: Thread[Finalizer,8,system]
THREAD 2: Thread[Reference Handler,10,system]
THREAD 3: Thread[Monitor Ctrl-Break,5,main]
THREAD 4: Thread[process reaper,10,system]
THREAD 5: Thread[Thread-103,5,main]
THREAD 6: Thread[ChangeTracker-http://localhost:4984/mobicat/,5,main]
THREAD 7: Thread[main,5,main]
End Test
```

--- a/src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java
+++ b/src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java
@@ -278,7 +278,8 @@ abstract class ReplicationInternal implements BlockingQueueListener{
     protected void close() {
         // shutdown ScheduledExecutorService. Without shutdown, cause thread leak
         if (remoteRequestExecutor != null && !remoteRequestExecutor.isShutdown()) {
-            Utils.shutdownAndAwaitTermination(remoteRequestExecutor);
+            // Note: Time to wait is set 60 sec because RemoteRequest's socket timeout is set 60 seconds.
+            Utils.shutdownAndAwaitTermination(remoteRequestExecutor, 60);
         }
     }
 

--- a/src/main/java/com/couchbase/lite/support/RemoteRequest.java
+++ b/src/main/java/com/couchbase/lite/support/RemoteRequest.java
@@ -25,7 +25,6 @@ import org.apache.http.client.methods.HttpPost;
 import org.apache.http.client.methods.HttpPut;
 import org.apache.http.client.methods.HttpUriRequest;
 import org.apache.http.client.protocol.ClientContext;
-import org.apache.http.conn.ClientConnectionManager;
 import org.apache.http.entity.ByteArrayEntity;
 import org.apache.http.impl.auth.BasicScheme;
 import org.apache.http.impl.client.DefaultHttpClient;
@@ -89,8 +88,6 @@ public class RemoteRequest implements Runnable {
             Log.v(Log.TAG_SYNC, "%s: RemoteRequest run() called, url: %s", this, url);
 
             HttpClient httpClient = clientFactory.getHttpClient();
-
-            ClientConnectionManager manager = httpClient.getConnectionManager();
 
             preemptivelySetAuthCredentials(httpClient);
 

--- a/src/main/java/com/couchbase/lite/support/RemoteRequestRetry.java
+++ b/src/main/java/com/couchbase/lite/support/RemoteRequestRetry.java
@@ -208,7 +208,12 @@ public class RemoteRequestRetry<T> implements Future<T> {
 
                 // Only retry if error is  TransientError (5xx).
                 if (isTransientError(httpResponse, e)) {
-                    if (retryCount >= MAX_RETRIES) {
+                    if (requestExecutor != null && requestExecutor.isShutdown()) {
+                        // requestExecutor was shutdown, no more retry.
+                        Log.e(Log.TAG_SYNC, "%s: RemoteRequestRetry failed, RequestExecutor was shutdown. url: %s", this, url);
+                        completed(httpResponse, result, e);
+                    }
+                    else if (retryCount >= MAX_RETRIES) {
                         Log.d(Log.TAG_SYNC, "%s: RemoteRequestRetry failed, but transient error.  retries exhausted. url: %s", this, url);
                         // ok, we're out of retries, propagate completion block call
                         completed(httpResponse, result, e);
@@ -299,16 +304,18 @@ public class RemoteRequestRetry<T> implements Future<T> {
 
         while (retryCount <= MAX_RETRIES) {
 
+            // requestExecutor was shutdown, no more retry.
+            if (requestExecutor == null || requestExecutor.isShutdown()) {
+                return null;
+            }
+
             // Take a future from the queue
             Future future = pendingRequests.take();
-
             future.get();
-
             if (completedSuccessfully.get() == true) {
                 // we're done
                 return null;
             }
-
         }
 
         // exhausted attempts, callback to original caller with result.  requestThrowable

--- a/src/main/java/com/couchbase/lite/util/Utils.java
+++ b/src/main/java/com/couchbase/lite/util/Utils.java
@@ -269,15 +269,17 @@ public class Utils {
      * first by calling shutdown to reject incoming tasks, and then calling shutdownNow,
      * if necessary, to cancel any lingering tasks:
      * http://docs.oracle.com/javase/6/docs/api/java/util/concurrent/ExecutorService.html
+     *
+     * @param timeToWait - Seconds
      */
-    public static void shutdownAndAwaitTermination(ExecutorService pool) {
+    public static void shutdownAndAwaitTermination(ExecutorService pool, long timeToWait) {
         pool.shutdown(); // Disable new tasks from being submitted
         try {
             // Wait a while for existing tasks to terminate
-            if (!pool.awaitTermination(20, TimeUnit.SECONDS)) {
+            if (!pool.awaitTermination(timeToWait, TimeUnit.SECONDS)) {
                 pool.shutdownNow(); // Cancel currently executing tasks
                 // Wait a while for tasks to respond to being cancelled
-                if (!pool.awaitTermination(20, TimeUnit.SECONDS)) {
+                if (!pool.awaitTermination(timeToWait, TimeUnit.SECONDS)) {
                     Log.e(Log.TAG_DATABASE, "Pool did not terminate");
                 }
             }
@@ -287,5 +289,9 @@ public class Utils {
             // Preserve interrupt status
             Thread.currentThread().interrupt();
         }
+    }
+
+    public static void shutdownAndAwaitTermination(ExecutorService pool) {
+        shutdownAndAwaitTermination(pool, 20);
     }
 }

